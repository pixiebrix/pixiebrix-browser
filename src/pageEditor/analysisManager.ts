/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import BlockTypeAnalysis from "@/analysis/analysisVisitors/blockTypeAnalysis";
import ExtensionUrlPatternAnalysis from "@/analysis/analysisVisitors/extensionUrlPatternAnalysis";
import OutputKeyAnalysis from "@/analysis/analysisVisitors/outputKeyAnalysis";
import RenderersAnalysis from "@/analysis/analysisVisitors/renderersAnalysis";
import TemplateAnalysis from "@/analysis/analysisVisitors/templateAnalysis";
import TraceAnalysis from "@/analysis/analysisVisitors/traceAnalysis";
import ReduxAnalysisManager from "@/analysis/ReduxAnalysisManager";
import { type UUID } from "@/core";
import { type TraceRecord } from "@/telemetry/trace";
import { isAnyOf, type PayloadAction } from "@reduxjs/toolkit";
import { type RootState } from "./pageEditorTypes";
import { actions as editorActions } from "@/pageEditor/slices/editorSlice";
import runtimeSlice from "./slices/runtimeSlice";
import RequestPermissionAnalysis from "@/analysis/analysisVisitors/requestPermissionAnalysis";
import FormBrickAnalysis from "@/analysis/analysisVisitors/formBrickAnalysis";
import { selectActiveElementTraces } from "./slices/runtimeSelectors";
import VarAnalysis from "@/analysis/analysisVisitors/varAnalysis/varAnalysis";
import analysisSlice from "@/analysis/analysisSlice";
import RegexAnalysis from "@/analysis/analysisVisitors/regexAnalysis";

const runtimeActions = runtimeSlice.actions;

const pageEditorAnalysisManager = new ReduxAnalysisManager();

// These actions will be used with every analysis so the annotation path is up-to-date
// with the node position in the pipeline
const nodeListMutationActions = [
  editorActions.addNode,
  editorActions.moveNode,
  editorActions.removeNode,
] as const;

// The order in which the analysis are registered is important.
// The first analysis registered will be the first to run.
// When multiple actions (e.g. typing) trigger analysis, the later analysis have more chances to get aborted.
// Try to put the faster analysis first, and the slower ones at the end.

pageEditorAnalysisManager.registerAnalysisEffect(
  (
    action: PayloadAction<{ extensionId: UUID; records: TraceRecord[] }>,
    state: RootState
  ) => {
    // TraceAnalysis filter the trace errors, thus
    // selecting all records here to avoid double filtering
    const records = selectActiveElementTraces(state);

    return new TraceAnalysis(records);
  },
  {
    matcher: isAnyOf(
      runtimeActions.setExtensionTrace,
      ...nodeListMutationActions
    ),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new BlockTypeAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new FormBrickAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new RenderersAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(() => new TemplateAnalysis(), {
  matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
});

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new ExtensionUrlPatternAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new RequestPermissionAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(() => new RegexAnalysis(), {
  matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
});

const varAnalysisFactory = (
  action: PayloadAction<{ extensionId: UUID; records: TraceRecord[] }>,
  state: RootState
) => {
  const records = selectActiveElementTraces(state);
  return new VarAnalysis(records);
};

// OutputKeyAnalysis seems to be the slowest one, so we register it in the end
pageEditorAnalysisManager.registerAnalysisEffect(
  () => new OutputKeyAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

// VarAnalysis is not the slowest itself, but it triggers a post-analysis action,
// so it is the last one
pageEditorAnalysisManager.registerAnalysisEffect(
  varAnalysisFactory,
  {
    matcher: isAnyOf(
      editorActions.editElement,
      runtimeActions.setExtensionTrace,
      ...nodeListMutationActions
    ),
  },
  {
    postAnalysisAction(analysis, extensionId, listenerApi) {
      listenerApi.dispatch(
        analysisSlice.actions.setKnownVars({
          extensionId,
          vars: analysis.getKnownVars(),
        })
      );
    },
  }
);

export default pageEditorAnalysisManager;
