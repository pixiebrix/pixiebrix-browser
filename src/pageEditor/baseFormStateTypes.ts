/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type Except } from "type-fest";
import {
  type IntegrationDependency,
  type IntegrationDependencyV1,
  type IntegrationDependencyV2,
} from "@/types/integrationTypes";
import { type ApiVersion, type OptionsArgs } from "@/types/runtimeTypes";
import { type UUID } from "@/types/stringTypes";
import { type StarterBrickType } from "@/starterBricks/types";
import { type Permissions } from "webextension-polyfill";
import { type ModComponentBase } from "@/types/modComponentTypes";
import { type ModOptionsDefinition } from "@/types/modDefinitionTypes";
import {
  type BrickPipeline,
  type NormalizedAvailability,
} from "@/bricks/types";
import { type Metadata, type RegistryId } from "@/types/registryTypes";

/**
 * A simplified type for ReaderConfig to prevent TypeScript reporting problems with infinite type instantiation
 * @see ReaderConfig
 */
export type SingleLayerReaderConfig =
  | RegistryId
  | RegistryId[]
  | Record<string, RegistryId>;

export type BaseExtensionPointState = {
  metadata: Metadata;
  definition: {
    type: StarterBrickType;
    // We're currently not allowing users to modify readers in the page editor
    reader: SingleLayerReaderConfig;
    isAvailable: NormalizedAvailability;
  };
};

export interface BaseExtensionState {
  blockPipeline: BrickPipeline;
}

/**
 * @deprecated - Do not use versioned state types directly
 */
export interface BaseFormStateV1<
  TExtension extends BaseExtensionState = BaseExtensionState,
  TExtensionPoint extends BaseExtensionPointState = BaseExtensionPointState
> {
  /**
   * The apiVersion of the brick definition, controlling how PixieBrix interprets brick definitions
   * @see ApiVersion
   */
  readonly apiVersion: ApiVersion;

  /**
   * The extension uuid
   */
  readonly uuid: UUID;

  /**
   * The type of the extensionPoint
   */
  readonly type: StarterBrickType;

  /**
   * True if the extensionPoint exists in the registry
   */
  installed?: boolean;

  /**
   * True if the extension should be allowed to auto-reload. In general, only extensions that require user
   * interaction to trigger should be allowed to auto-reload. Otherwise, PixieBrix might end up spamming an API
   */
  autoReload?: boolean;

  /**
   * User-provided name to identify the extension
   */
  label: string;

  /**
   * The input options from the extension's blueprint
   * @since 1.4.3
   */
  optionsArgs: OptionsArgs;

  services: IntegrationDependencyV1[];

  /**
   * The extra permissions required by the extension
   * @since 1.7.0
   */
  permissions: Permissions.Permissions;

  extensionPoint: TExtensionPoint;

  extension: TExtension;

  /**
   * Information about the recipe (i.e., blueprint) used to install the extension, or `undefined` if the extension
   * is not part of a recipe.
   * @see ModComponentBase._recipe
   */
  recipe: ModComponentBase["_recipe"] | undefined;

  /**
   * Information about the recipe (i.e., blueprint) options,
   * or `undefined` if the extension is not part of a recipe.
   * @see ModDefinition.options
   */
  optionsDefinition?: ModOptionsDefinition;
}

/**
 * @deprecated - Do not use versioned state types directly
 */
export type BaseFormStateV2<
  TExtension extends BaseExtensionState = BaseExtensionState,
  TExtensionPoint extends BaseExtensionPointState = BaseExtensionPointState
> = Except<BaseFormStateV1<TExtension, TExtensionPoint>, "services"> & {
  /**
   * The integration dependencies configured for the extension
   *
   * @since 1.7.41 renamed from `services` to `integrationDependencies`, also
   * changed from IntegrationDependencyV1 to IntegrationDependencyV2
   */
  integrationDependencies: IntegrationDependencyV2[];
};
export type BaseFormState<
  TExtension extends BaseExtensionState = BaseExtensionState,
  TExtensionPoint extends BaseExtensionPointState = BaseExtensionPointState
> = Except<
  BaseFormStateV2<TExtension, TExtensionPoint>,
  "integrationDependencies"
> & {
  /**
   * Using the un-versioned type
   */
  integrationDependencies: IntegrationDependency[];
};
