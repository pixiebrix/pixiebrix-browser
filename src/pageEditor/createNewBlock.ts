/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { BlockConfig } from "@/blocks/types";
import { defaultBlockConfig } from "@/blocks/util";
import { RegistryId, Schema } from "@/core";
import { uuidv4 } from "@/types/helpers";
import { getExampleBlockConfig } from "./exampleBlockConfigs";

export function createNewBlock(
  blockId: RegistryId,
  blockInputSchema?: Schema
): BlockConfig {
  return {
    id: blockId,
    instanceId: uuidv4(),
    config:
      getExampleBlockConfig(blockId) ??
      (blockInputSchema == null ? {} : defaultBlockConfig(blockInputSchema)),
  };
}
