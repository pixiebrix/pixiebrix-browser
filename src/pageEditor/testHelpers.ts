/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import {
  createRenderHookWithWrappers,
  createRenderWithWrappers,
} from "@/testUtils/testHelpers";
import analysisSlice from "@/analysis/analysisSlice";
import pageEditorAnalysisManager from "./analysisManager";
import { tabStateSlice } from "@/pageEditor/tabState/tabStateSlice";
import { appApi } from "@/services/api";
import { modDefinitionsSlice } from "@/modDefinitions/modDefinitionsSlice";
import { modDefinitionsMiddleware } from "@/modDefinitions/modDefinitionsListenerMiddleware";

const configureStoreForTests = () =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      analysis: analysisSlice.reducer,
      tabState: tabStateSlice.reducer,
      modDefinitions: modDefinitionsSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
    middleware(getDefaultMiddleware) {
      /* eslint-disable unicorn/prefer-spread -- It's not Array#concat, can't use spread */
      return getDefaultMiddleware()
        .concat(appApi.middleware)
        .concat(pageEditorAnalysisManager.middleware)
        .concat(modDefinitionsMiddleware);
      /* eslint-enable unicorn/prefer-spread */
    },
  });

const renderWithWrappers = createRenderWithWrappers(configureStoreForTests);
const renderHookWithWrappers = createRenderHookWithWrappers(
  configureStoreForTests
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
export { renderHookWithWrappers as renderHook };
