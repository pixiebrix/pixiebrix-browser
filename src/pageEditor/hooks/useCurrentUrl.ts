/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { once } from "lodash";
import { useEffect, useState } from "react";
import { type Target } from "@/types";
import { SimpleEventTarget } from "@/utils/SimpleEventTarget";
import { type WebNavigation } from "webextension-polyfill";
import { expectContext } from "@/utils/expectContext";
import { getCurrentURL } from "@/pageEditor/utils";

let tabUrl: string;
const TOP_LEVEL_FRAME_ID = 0;

const urlChanges = new SimpleEventTarget<string>();

// The pageEditor only cares for the top frame
function isCurrentTopFrame({ tabId, frameId }: Target) {
  return (
    frameId === TOP_LEVEL_FRAME_ID &&
    tabId === browser.devtools.inspectedWindow.tabId
  );
}

async function onNavigation(
  target: WebNavigation.OnCommittedDetailsType
): Promise<void> {
  if (isCurrentTopFrame(target)) {
    tabUrl = target.url;
    urlChanges.emit(target.url);
  }
}

const startWatching = once(async () => {
  browser.webNavigation.onCommitted.addListener(onNavigation);
  tabUrl = await getCurrentURL();
  urlChanges.emit(tabUrl);
});

export default function useCurrentUrl(): string {
  expectContext("devTools");

  const [url, setUrl] = useState(tabUrl);

  useEffect(() => {
    urlChanges.add(setUrl);
    void startWatching();
    return () => {
      urlChanges.remove(setUrl);
    };
  }, [setUrl]);

  return url;
}
