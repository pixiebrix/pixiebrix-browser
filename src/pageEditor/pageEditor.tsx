/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/bootstrapWithoutRem.css";
import "@/utils/global.scss";
import "@/utils/layout.scss";

import "@/extensionContext";
import "@/development/darkMode.js";

import { initMessengerLogging } from "@/development/messengerLogging";

import ReactDOM from "react-dom";
import React from "react";
import Panel from "@/pageEditor/Panel";
import { watchNavigation } from "@/pageEditor/protocol";
import { initToaster } from "@/utils/notify";
import { initRuntimeLogging } from "@/development/runtimeLogging";
import { initPerformanceMonitoring } from "@/telemetry/performance";
import { setPlatform } from "@/platform/platformContext";
import extensionPagePlatform from "@/extensionPages/extensionPagePlatform";

async function init() {
  setPlatform(extensionPagePlatform);
  void initMessengerLogging();
  void initRuntimeLogging();
  try {
    await initPerformanceMonitoring();
  } catch (error) {
    console.error("Failed to initialize performance monitoring", error);
  }

  watchNavigation();
  initToaster();

  ReactDOM.render(<Panel />, document.querySelector("#container"));
}

void init();
