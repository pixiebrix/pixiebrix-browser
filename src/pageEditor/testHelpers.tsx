/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import { appApi } from "@/services/api";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
