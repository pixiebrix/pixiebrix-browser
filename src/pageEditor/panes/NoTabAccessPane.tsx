/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import Centered from "@/components/Centered";
import { parse as parseDomain } from "psl";
import useCurrentUrl from "@/pageEditor/hooks/useCurrentUrl";
import { useSelector } from "react-redux";
import { selectIsContextInvalidated } from "@/pageEditor/tabState/tabStateSelectors";
import { CONTEXT_INVALIDATED_MESSAGE } from "@/pageEditor/tabState/tabStateSlice";
import { safeParseUrl } from "@/utils/urlUtils";

function getPageLabel(url: string): string {
  const { hostname } = safeParseUrl(url);
  const result = parseDomain(hostname);
  if ("domain" in result && result.domain) {
    return result.domain;
  }
}

/**
 * Panel the shows that PixieBrix does not have access to the page. Now that all URLs are included in the manifest's
 * required permissions, this can occur in three cases:
 * 1. The browser extension was restarted while page was open
 * 2. The extension context was invalidated in tab
 * 3. Permissions are blocked via runtime_blocked_hosts
 * @constructor
 */
const NoTabAccessPane: React.FunctionComponent = () => {
  const url = useCurrentUrl();
  const siteLabel = (url && getPageLabel(url)) || "this page";

  const isContextInvalidated = useSelector(selectIsContextInvalidated);

  if (isContextInvalidated) {
    // Ideally, if the context was invalidated, we'd also provide a button to automatically reload the inspected tab.
    // However, because the context is invalidated, we can't send the message to reload it.
    return (
      <Centered vertically>
        <p>{CONTEXT_INVALIDATED_MESSAGE}</p>
      </Centered>
    );
  }

  return (
    <Centered vertically>
      <p className="font-weight-bold">
        PixieBrix cannot connect to {siteLabel}. Try reloading the Tab.
      </p>

      <p className="text-muted">
        If PixieBrix is managed by your Enterprise, contact your IT
        Administrator to ensure this URL is allowed by your {"organization's"}{" "}
        Chrome Extension Policy.
      </p>
    </Centered>
  );
};

export default NoTabAccessPane;
