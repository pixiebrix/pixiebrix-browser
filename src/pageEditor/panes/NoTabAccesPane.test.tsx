/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import NoTabAccessPane from "@/pageEditor/panes/NoTabAccessPane";
import { render, screen } from "@/pageEditor/testHelpers";
import useCurrentUrl from "@/pageEditor/hooks/useCurrentUrl";
import { waitFor } from "@testing-library/react";

jest.mock("@/pageEditor/hooks/useCurrentUrl", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("PermissionsPane", () => {
  test("it renders", () => {
    (useCurrentUrl as jest.Mock).mockReturnValue("https://test.url");
    const rendered = render(<NoTabAccessPane />);
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  test("it renders right copy when the URL isn't available", async () => {
    (useCurrentUrl as jest.Mock).mockReturnValue(undefined);
    render(<NoTabAccessPane />);

    await waitFor(() => {
      expect(
        screen.getByText(
          "PixieBrix cannot connect to this page. Try reloading the Tab."
        )
      ).toBeInTheDocument();
    });
  });
});
