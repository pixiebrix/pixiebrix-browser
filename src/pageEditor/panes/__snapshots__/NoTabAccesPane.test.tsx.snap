// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionsPane it renders 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column mx-auto mt-4 pb-2 max-550 text-center h-100 justify-content-center"
  >
    <p>
      PixieBrix cannot connect to test.url. Please try reloading the Tab.
    </p>
    <p>
      If you are part of an Enterprise deployment, please contact your IT Administrator to ensure this URL is allowed by your organization's Chrome Extension Policy,
    </p>
  </div>
</DocumentFragment>
`;
